### ISSUE TITLE
FMDatabaseQueue's use of serial dispatch queue not enough to guarantee thread safety

### ISSUE THREAD

**Issue by fumoboy007** [ccgus/fmdb#131]
I'm interfacing with a SQLite database using one `FMDatabaseQueue`. On occasion, I get an `EXC_BAD_ACCESS` from one of the SQLite functions. If I put the following in FMDatabase.m, I don't get them anymore:

```
+ (void)initialize {
    if (sqlite3_shutdown() != SQLITE_OK)
        NSLog(@"Couldn't shut down SQLite.");

    if (sqlite3_config(SQLITE_CONFIG_SERIALIZED) != SQLITE_OK)
        NSLog(@"Couldn't set serialized mode.");

    if (sqlite3_initialize() != SQLITE_OK)
        NSLog(@"Couldn't initialize SQLite.");
}
```

So clearly the `EXC_BAD_ACCESS` exceptions have something to do with multithreading. My theory is that one/some of the SQLite functions execute asynchronously and because a serial dispatch queue may run on different threads (even though it executes tasks serially), a race condition occurs.

In conclusion, `FMDatabaseQueue` needs to be rewritten to guarantee the use of only one thread. I may undertake this if I have time. ;)

**Comment by iNeerajatSigneasy** [comment]
Hi I am facing the same issue of getting random crashes on EXC_BAD_ACCESS. I can see 2 threads having access to DB in my stack trace. I use a singleton DBQueue to manage Read, Write.  Although I am also not closing resultsets because previous developer was idiot enough to save resultsets and serve data to view instead of keeping our own models. I don't have time for revamp and I am trying to understand what else can cause this?

**Comment by ccgus** [comment]
Can you paste the stack traces somewhere?

**Comment by Amitmundra** [comment]
I am also facing BAD acess issue, can you please help me, how to troubleshoot issue?

**Comment by robertmryan** [comment]
ARC or manual reference counting? If manual reference counting, did you run it through the static analyzer ("Analyze" on Xcode's "Product" menu)? You should have a clean bill of health from that before we go further. 

If that's not it, we need the smallest possible reproducible example of your problem (i.e. we don't want to wade through all of your code, but we need absolute minimum necessary to reproduce the problem).

**Comment by robertmryan** [comment]
It looks like this is moved to issue https://github.com/ccgus/fmdb/issues/404, which is good, because this doesn't belong here in this closed discussion.

**Comment by Amitmundra** [comment]
I am using ARC reference counting.
Do you know any reason of crashes, some time it crashes while lopping FMResultSet?
Please help and give any suggestion.

Thanks.
